### PR TITLE
[C++ Verification] avoid the temporary object

### PR DESCRIPTION
**Pull request by XLiZHI**
This PR attempts to avoid some temporary objects generated by member initialization.

Try to resolve the issue:
- https://github.com/esbmc/esbmc/issues/1729

Fixed 4 TCs, ~~caused 1 KOWNBUG:~~

~~`regression/esbmc-cpp/cpp/ch21-7` : dereference failure: invalid pointer, this TC was fixed in the last PR and seems to be caused by temporary obj.~~

It is not clear to me at this time why temporary variables affect constructors, it seems to cause some variables to fail to initialize. It would be nice to explain what temporary objects do.

### CODE REVIEW

**Review by lucasccordeiro — APPROVED** [review]

**Comment by fbrausse** [comment]
> It is not clear to me at this time why temporary variables affect constructors, it seems to cause some variables to fail to initialize. It would be nice to explain what temporary objects do.

They are handled as goto-sideeffects and introduce a new symbol (with a new name). The tmp-obj expression itself is being replaced with a symbol_exprt referring to this new symbol. This happens in goto_convertt::remove_temporary_object(). These tmp-obj expressions have two optional attributes:
1. If the tmp-obj expression has an operand, it is used as the initialization of the new symbol (via assignment).
2. If the tmp-obj expression has an .initializer() (which is supposed to contain code that can do arbitrary things with the new symbol), then all `new_object` expressions in that initializer will get replaced by the symbol_expr referring to the new symbol, see goto_convertt::replace_new_object().

The initializer and  `new_object` expression being used in get_constructor_call() is basically a reference to the anonymous symbol, which the `else` branch was using to build the `this` pointer, for instance when there is no declaration attached to the constructor call. E.g. something like a temporary here:
```c++
struct A { int x; A(int x) : x(x) {} };

A a(A(42).x);
```

~Now that I look at this, I'm wondering whether the `new_object` expression, which is assigned to the operand of the tmp-obj expression (i.e., case 1) in the `else` branch of get_constructor_call() is actually being replaced by the proper symbol_expr.~

*Edit:* Sorry, my bad, it's using case 2. That should be fine.

**Comment by XLiZHI** [comment]
> They are handled as goto-sideeffects and introduce a new symbol (with a new name). The tmp-obj expression itself is being replaced with a symbol_exprt referring to this new symbol. This happens in goto_convertt::remove_temporary_object(). These tmp-obj expressions have two optional attributes:
> 
> 1. If the tmp-obj expression has an operand, it is used as the initialization of the new symbol (via assignment).
> 2. If the tmp-obj expression has an .initializer() (which is supposed to contain code that can do arbitrary things with the new symbol), then all `new_object` expressions in that initializer will get replaced by the symbol_expr referring to the new symbol, see goto_convertt::replace_new_object().

Thanks @fbrausse, I've read this code before, and what makes it difficult for me is that we do not have enough information to determine whether temporary objects are needed. I've been working on it for a few days, but I haven't made any progress. In addition, we really need some simple TCs to explore this problem. 

> Now that I look at this, I'm wondering whether the new_object expression, which is assigned to the operand of the tmp-obj expression (i.e., case 1) in the else branch of get_constructor_call() is actually being replaced by the proper symbol_expr.

At present, I use this PR to avoid some errors temporarily, but I think we should try to avoid temporary objects, which will make the goto program very large. What else do you think we need to do in this PR?

**Comment by fbrausse** [comment]
> At present, I use this PR to avoid some errors temporarily, but I think we should try to avoid temporary objects, which will make the goto program very large. What else do you think we need to do in this PR?

1. It would of course be good to find a more permanent solution, maybe in this PR.
2. We can't always avoid temporary objects, e.g. in the example above: we need to call the ctor of `A` twice there, once with the declared symbol `a` and once with a temporary. Whether this is to be done via a `temporary_object` expression or by introducing a new symbol in the converter directly (as done for compound literals, for instance), I don't know. Ideally, the tmp-obj expression would do what it's supposed to.

For compound literals, see https://github.com/esbmc/esbmc/blob/2f5a0ee8f81ad27a4b2fa5956213e8dcc2d56eee/src/clang-c-frontend/clang_c_convert.cpp#L1732 and https://github.com/esbmc/esbmc/blob/2f5a0ee8f81ad27a4b2fa5956213e8dcc2d56eee/src/clang-c-frontend/clang_c_convert.cpp#L531

I tried reusing this code in my last PR about C++ references, but it basically inserted the symbol declarations at the wrong position in the code that's why I stayed with the tmp-obj expression there.

@kunjsong01 did call need_new_object() in that place before my last change: https://github.com/esbmc/esbmc/blob/2f5a0ee8f81ad27a4b2fa5956213e8dcc2d56eee/src/clang-cpp-frontend/clang_cpp_convert.cpp#L819-L824

However, as it just examines the direct parent expression, this didn't work: there could be casts and other things in the parent expression(s).

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 1]
```diff
@@ -826,8 +826,15 @@ bool clang_cpp_convertert::get_constructor_call(
    */
 
   // Calling base constructor from derived constructor
+  bool need_new_obj = false;
+
   if (new_expr.base_ctor_derived())
     gen_typecast_base_ctor_call(callee_decl, call, new_expr);
+  else if (
+    new_expr.get_bool("memberInit") &&
```
Comment (fbrausse) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: Let's stick to the non-capitalized names delimited by underscores. Also, I believe, this should be prefixed by `#` in order to indicate that it's a "comment".

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 2]
```diff
@@ -826,8 +826,15 @@ bool clang_cpp_convertert::get_constructor_call(
...
+    !callee_decl.type().return_type().get_bool("#copy_cons"))
```
Comment (fbrausse) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: Could you add a comment indicating why copy-constructors are special here? For instance, I'm wondering what move-constructors would do, or actually in what way user-defined constructors are different.

**Comment by XLiZHI** [comment]
```
        A a;
        // 23 file main8.cpp line 7 column 5 function main
        A tmp$1;
        // 24 file main8.cpp line 7 column 5 function main
        tmp$1=NONDET(struct);
        // 25 file main8.cpp line 7 column 7 function main
        A tmp$2;
        // 26 file main8.cpp line 7 column 7 function main
        tmp$2=NONDET(struct);
        // 27 
        FUNCTION_CALL:  A(&tmp$2, 42)
        // 28 
        FUNCTION_CALL:  A(&tmp$1, tmp$2.x)
        // 29 file main8.cpp line 7 column 3 function main
        a=tmp$1;
```

```
`-FunctionDecl 0x82bad58 <line:5:1, line:9:1> line:5:5 main 'int ()'
  `-CompoundStmt 0x82bb2c0 <line:6:1, line:9:1>
    `-DeclStmt 0x82bb2a8 <line:7:3, col:15>
      `-VarDecl 0x82bae40 <col:3, col:14> col:5 a 'A' callinit
        `-CXXConstructExpr 0x82bb278 <col:5, col:14> 'A' 'void (int)'
          `-MemberExpr 0x82bb228 <col:7, col:13> 'int' .x 0x82b4858
            `-CXXFunctionalCastExpr 0x82bb200 <col:7, col:11> 'A' functional cast to struct A <ConstructorConversion>
              `-CXXConstructExpr 0x82bb0b0 <col:7, col:11> 'A' 'void (int)'
                `-IntegerLiteral 0x82baeb8 <col:9> 'int' 42
```

> However, as it just examines the direct parent expression, this didn't work: there could be casts and other things in the parent expression(s).

Actually I have tried, but I don't think we can stably use AST, as you said the parent nodes may exist cast.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 2 above)
Comment (XLiZHI) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp]: After further investigation, I found that the arguments of the cpy constructor was wrong:
```
#include <cassert>

struct B {
    int data;

    B(int x) : data(x) {}
};

struct A {
    B b;

    A(const B& bCopy) : b(bCopy) {}
};

int main() {
    B bOriginal(10);
    A a(bOriginal);

    assert(a.b.data == 10);

    return 0;
}
```
use tmp obj:

```
A (c:@S@A@F@A#&1$@S@B#):
        // 24 file main10.cpp line 12 column 25 function A
        B tmp$1;
        // 25 file main10.cpp line 12 column 25 function A
        tmp$1=NONDET(struct);
        // 26 
        FUNCTION_CALL:  B(&tmp$1, &(*bCopy))
        // 27 
        this->b=tmp$1;
        // 28 file main10.cpp line 12 column 35 function A
        dead tmp$1;
        // 29 file main10.cpp line 12 column 35 function A
        END_FUNCTION // A
```
original:
```
A (c:@S@A@F@A#&1$@S@B#):
        // 24
        FUNCTION_CALL:  B(&(*bCopy), &this->b)   // should be B( &this->b, &(*bCopy)) I think.
        // 25 file main10.cpp line 14 column 35 function A
        END_FUNCTION // A
```

I haven't tried move constructor yet, but I think it should be the same.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 3]
```diff
@@ -826,8 +826,17 @@ bool clang_cpp_convertert::get_constructor_call(
...
   // Calling base constructor from derived constructor
+  bool need_new_obj = false;
+
   if (new_expr.base_ctor_derived())
     gen_typecast_base_ctor_call(callee_decl, call, new_expr);
+  else if (new_expr.get_bool("#member_init"))
+  {
+    /*
+    struct A {}; struct B {A a; B(int _a) : a(_a) {}};
+    In this case, use variables to construct the object directly
```
Comment (fbrausse) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: `A` needs a constructor taking an int. Also, I'd replace "variables" by "members" for clarity.

**Review by fbrausse — APPROVED** [review]
Nice, good catch with the adjuster call! I missed that. LGTM! One nitpick.

**Pull request merged**